File: src/111.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tree.h"

#define DFS (1)
#define BFS (1)
#if (DFS)
int dfs(struct TreeNode* root) {
    int retVal = 0;

    if (root == NULL) {
        return retVal;
    }

    if ((root->left == NULL) && (root->right == NULL)) {
        retVal += 1;
        return retVal;
    }

    if (root->left == NULL) {
        retVal = 1 + dfs(root->right);
    } else if (root->right == NULL) {
        retVal = 1 + dfs(root->left);
    } else {
        retVal = 1 + fmin(dfs(root->left), dfs(root->right));
    }

    return retVal;
}
#elif (BFS)
#define MAX_QUEUE_SIZE (100000 * 2)  // The number of nodes in the tree is in the range [0, 10^5].
int bfs(struct TreeNode* root) {
    int retVal = 0;

    int bfsQueueHead = 0;
    int bfsQueueTail = 0;
    struct TreeNode* bfsQueue[MAX_QUEUE_SIZE];
    bfsQueue[bfsQueueTail] = root;
    ++bfsQueueTail;
    retVal = 1;

    int bfsQueueSize;
    struct TreeNode* pCurrent;
    int i;
    while (bfsQueueHead < bfsQueueTail) {
        bfsQueueSize = bfsQueueTail - bfsQueueHead;
        for (i = 0; i < bfsQueueSize; ++i) {
            pCurrent = bfsQueue[bfsQueueHead];
            ++bfsQueueHead;

            if ((pCurrent->left == NULL) && (pCurrent->right == NULL)) {
                return retVal;
            }

            if (pCurrent->left != NULL) {
                bfsQueue[bfsQueueTail] = pCurrent->left;
                ++bfsQueueTail;
            }
            if (pCurrent->right != NULL) {
                bfsQueue[bfsQueueTail] = pCurrent->right;
                ++bfsQueueTail;
            }
        }
        ++retVal;
    }

    return retVal;
}
#endif
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int minDepth(struct TreeNode* root) {
    int retVal = 0;

    if (root == NULL) {
        return retVal;
    }

#if (DFS)
    printf("Depth-First Search\n\n");
    retVal = dfs(root);
#elif (BFS)
    printf("Breadth-First Search\n\n");
    retVal = bfs(root);
#endif

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (10000)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
    } testCase[] = {{{3, 9, 20, INT_MIN, INT_MIN, 15, 7}, 7},
                    {{2,       INT_MIN, 3,       INT_MAX, INT_MAX, INT_MIN, 4,       INT_MAX, INT_MAX, INT_MAX, INT_MAX,
                      INT_MAX, INT_MAX, INT_MIN, 5,       INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX,
                      INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MAX, INT_MIN, 6},
                     31}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: root = [3,9,20,null,null,15,7]
     *  Output: 2
     *
     *  Input: root = [2,null,3,null,4,null,5,null,6]
     *  Output: 5
     */

    struct TreeNode* pRoot = NULL;
    int answer = 0;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: root = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            if (testCase[i].nums[j] == INT_MIN) {
                printf("%snull", (j == 0 ? "" : ","));
            } else if (testCase[i].nums[j] == INT_MAX) {
                continue;
            } else {
                printf("%s%d", (j == 0 ? "" : ","), testCase[i].nums[j]);
            }
        }
        printf("]\n");
        pRoot = buildBinaryTree(testCase[i].nums, testCase[i].numsSize, 0);

        answer = minDepth(pRoot);
        printf("Output: %d\n", answer);

        printf("\n");

        freeTreeNode(pRoot);
        pRoot = NULL;
    }

    return EXIT_SUCCESS;
}

File: src/111.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from tree import TreeNode


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.method = 1

        if self.method == 1:
            print("Depth-First Search")
        elif self.method == 2:
            print("Breadth-First Search")
        print()

    def dfs(self, root: Optional[TreeNode]) -> int:
        retVal = 0

        if root == None:
            return retVal

        if (root.left == None) and (root.right == None):
            retVal += 1
            return retVal

        if root.left == None:
            retVal = 1 + self.dfs(root.right)
        elif root.right == None:
            retVal = 1 + self.dfs(root.left)
        else:
            retVal = 1 + min(self.dfs(root.left), self.dfs(root.right))

        return retVal

    def bfs(self, root: Optional[TreeNode]) -> int:
        retVal = 0

        bfsQueue = deque()
        bfsQueue.append(root)
        retVal = 1
        while bfsQueue:
            for _ in range(len(bfsQueue)):
                node = bfsQueue.popleft()

                if (node.left == None) and (node.right == None):
                    return retVal

                if node.left != None:
                    bfsQueue.append(node.left)
                if node.right != None:
                    bfsQueue.append(node.right)

            retVal += 1

        return retVal

    def minDepth(self, root: Optional[TreeNode]) -> int:
        retVal = 0

        if root == None:
            return retVal

        if self.method == 1:
            retVal = self.dfs(root)
        elif self.method == 2:
            retVal = self.bfs(root)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for root in [[3, 9, 20, None, None, 15, 7], [2, None, 3, None, 4, None, 5, None, 6]]:
            # /* Example
            #  *  Input: root = [3,9,20,null,null,15,7]
            #  *  Output: 2
            #  *
            #  *  Input: root = [2,null,3,null,4,null,5,null,6]
            #  *  Output: 5
            #  */
            logging.info("Input: root = %s", root)

            retVal = pSolution.minDepth(TreeNode().buildTree(root))
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
